Utilities

eboyda opened the pull request:
Merge rebuild.py into master.

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,156 @@
+    labels."""
+    # WIP?  It seems to work but returns slightly different
+    # numbers on MNIST than seen during training: On val set: .0743 vs
+    # .075 during training; on test set: .0775 vs .0749 during training.
```
Have you tried 0.0765432?

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,156 @@
+
+def save_params(file_prefix,params):
+    """Save weights and biases for all NN layers, one layer per file."""
+    for layer in range(len(params)/2):
```
is this `2` a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming))? If so, better to pull it out into its own variable.

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,156 @@
+    """Save weights and biases for all NN layers, one layer per file."""
+    for layer in range(len(params)/2):
+        outfile = file_prefix+'Layer'+str(layer)+'.pkl'
+        save_layer(outfile,*params[2*layer:2*(layer+1)])
```
ditto re: magic numbers

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,156 @@
+    """Load pickled weights and biases for a single NN layer.  Assumes weights
+    and biases were pickled in sequence into the specified file.
+    """
+    W = theano.shared(
```
I assume `W` stands for weights? I'm generally not a fan of one-letter variable names but there are legit reasons, including here probably. So I thought I'd flag this and `b`. Here's a quick comment on why this jumped out at me:

http://programmers.stackexchange.com/a/71720

    A short variable name which represents its intent clearly is to be preferred over a big variable name.

So if you've got small scope of usage, or large scope with clear intent, then it's all good. It's kind of a religious thing though - there's no right answer. No need to change.

matthewhanson replied:
Even in the case where a single letter accurately describes something (e.g, using same names as in well known equations such as c for speed of light), all it takes is one attempt to grep for it in a tree of files to see it's not a great idea no matter what.

eboyda commented:
Trial by fire!  I renamed the offending variables, rewrote the function with potential magic number to be more pythonic, and improved formatting per python style guide.  Check it out.  Thanks, guys.

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,155 @@
+    return numpy.argmax(fire_neurons(inputvecs,output_neurons,net),axis=1)
+
+
+def logreg_validate(input,labels,classifier):
```
looks like pure `input` is still around ...

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,155 @@
+    loss = theano.function(
+        inputs=[],
+        outputs=classifier.errors(labels),
+        givens={classifier.input:input}
```
here too

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,155 @@
+
+### REBUILD ###
+
+def load_params(file):
```
These `file` variables need to be renamed too.

robinkraft commented on code/rebuild.py:
```diff
@@ -0,0 +1,155 @@
+        name='b',
+        borrow=True
+    )
+    with open(file,'rb') as inf:
```
here.

robinkraft commented:
@eboyda I'm still seeing some random `W` and `b` variables around. Is that intentional? I just want to be sure they're not just slipping through the cracks.

eboyda commented:
Robin, I'm using W and b to comport with theano conventions, though they use them as such only as class attributes (i.e. self.W, self.b).  When they are simple variables would you prefer that I call them weights and biases instead?

robinkraft commented:
Naw it's all good - if there's a reason for the names then that's cool.

Just be sure to fix up the remaining `file` and `input` variables.

eboyda commented:
Thanks, and will do.  I think I lost some edits switching git branches... but I'm getting the hang of it.

robinkraft commented:
Bummer about the edits ... `git stash` may be your new best friend. Let's you temporarily set aside edits and get them back.

http://www.gitguys.com/topics/temporarily-stashing-your-work/

eboyda commented:
Cool tip with git stash, thx.  I was surprised to learn that I can have files from two different branches open at the same time, but saving them doesn't necessarily work!

robinkraft commented:
Merging. I'll bug you about tests another time :)
